packages/dds/shared-object-base/src/sharedObject.ts

```diff
@@ -186,7 +199,36 @@ export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedO
     /**
      * {@inheritDoc (ISharedObject:interface).snapshot}
      */
-    public abstract snapshot(): ITree;
+    public summarize(fullTree: boolean = false): IChannelSummarizeResult {
+        // Create a SummarySerializer that will be used to serialize any IFluidHandles in this object. The
+        // IFluidHandles represents route to any referenced Fluid object.
+        // SummarySerializer tracks the routes of all handles that it serializes.
+        const serializer = new SummarySerializer(this.runtime.channelsRoutingContext);
+
+        const snapshot: ITree = this.snapshotCore(serializer);
```

Comment:
I do not know if it's good idea long term (as it assumes very specific workflow of today), but I worry about mistakes in the code using wrong serializer. With given structure, what we can do to catch those cases is
1. Add boolean flag in this class that indicates that are we in summarize() method.
2. make this.seriqlizer a getter and assert in it that it is not being called while # 1 is true.

Reason it might be bad idea:
1. It assumes that snapshot() code will not result in loading anything. This comes from it being synchronous today, but in future it might be different. It's possible to address it by reducing assertion to serializing using "global" serializer, and allowing deserialization to happen just fine.

Reply:
I added the boolean flag.
This is not how the code will look long term. We will probably have to come up with stronger guarantees that we get all the handles during the summarize process. Currently, this also assumes that a DDS will serialize all its handles as part of summarize which also may not be true in the future.

I think this is fine for how things currently are. We will definitely have to revisit this once an initial (prototype) GC implementation is checked in.